### PR TITLE
[XItemStack] Major Changes - Map, Book & SpawnEgg Support

### PR DESCRIPTION
**Pull request by DeadSilenceIV**
Hello i added the BookMeta and the MapMeta into the project, also the SkullMeta serialization and deserialization was fixed for legacy versions. I also fixed some unsupported methods for legacy versions.

### CODE REVIEW

**Comment by DeadSilenceIV** [comment]
Oh i and i forgot that i also added the SpawnEggMeta.

File: src/main/java/com/cryptomorin/xseries/XItemStack.java [hunk 1]
```diff
@@ -88,28 +91,39 @@ public static void serialize(@Nonnull ItemStack item, @Nonnull ConfigurationSect
...
                 if (damageable.hasDamage()) config.set("damage", damageable.getDamage());
             }
-        } else if (item.getType().getMaxDurability() > 0) {
+        }else{
             config.set("damage", item.getDurability());
         }
-        config.set("material", item.getType().name());
+
+        // Display Name & Lore
+        if (meta.hasDisplayName())config.set("name", meta.getDisplayName());
```
Comment (CryptoMorin) [inline src/main/java/com/cryptomorin/xseries/XItemStack.java:113]: Why did you move this and removed `ChatColor.stripColor` from it?

File: src/main/java/com/cryptomorin/xseries/XItemStack.java [hunk 2]
```diff
@@ -404,7 +467,7 @@ public static ItemStack deserialize(@Nonnull ConfigurationSection config) {
                     builder.trail(fw.getBoolean("trail"));
                     builder.with(Enums.getIfPresent(FireworkEffect.Type.class, fw.getString("type").toUpperCase(Locale.ENGLISH)).or(FireworkEffect.Type.STAR));
 
-                    List<String> fwColors = fw.getStringList("colors");
+                    List<String> fwColors = fw.getStringList("base-colors");
```
Comment (CryptoMorin) [inline src/main/java/com/cryptomorin/xseries/XItemStack.java:470]: Why?

File: src/main/java/com/cryptomorin/xseries/XMaterial.java [hunk 3]
```diff
@@ -1575,18 +1575,19 @@ public static XMaterial matchXMaterial(@Nonnull ItemStack item) {
...
             if (xMaterial.isPresent()) return xMaterial;
         }
-        // Usually flat versions wouldn't pass this point, but some special materials do.
 
+        // Usually flat versions wouldn't pass this point, but some special materials do.
         XMaterial oldXMaterial = requestOldXMaterial(name, data);
         if (oldXMaterial == null) {
             // Special case. Refer to FILLED_MAP for more info.
-            return data > 0 && name.endsWith("MAP") ? Optional.of(FILLED_MAP) : Optional.empty();
+            return (data >= 0 && isAMap) ? Optional.of(FILLED_MAP) : Optional.empty();
```
Comment (CryptoMorin) [inline src/main/java/com/cryptomorin/xseries/XMaterial.java:1590]: Please also explain all these changes in this file.

File: src/main/java/com/cryptomorin/xseries/XItemStack.java [hunk 4]
```diff
@@ -226,8 +239,43 @@ public static void serialize(@Nonnull ItemStack item, @Nonnull ConfigurationSect
                 fwc.set("fade-colors", fadeColors);
                 i++;
             }
-            //} else if (meta instanceof MapMeta) {
-        } else if (XMaterial.supports(14)) {
+        } else if(meta instanceof BookMeta) {
+            BookMeta book = (BookMeta)meta;
+            ConfigurationSection info = config.createSection("information");
```
Comment (CryptoMorin) [inline src/main/java/com/cryptomorin/xseries/XItemStack.java:244]: I don't think they necessarily need to be in an config section. Why not just add `title`, `author` and etc... in the main section?

File: src/main/java/com/cryptomorin/xseries/XItemStack.java [hunk 5]
```diff
@@ -226,8 +239,43 @@ public static void serialize(@Nonnull ItemStack item, @Nonnull ConfigurationSect
...
+                BookMeta.Generation generation = book.getGeneration();
+                if (generation != null) {
+                    info.set("generation", book.getGeneration().toString());
+                }
+            }
+            info.set("pages", book.getPages());
+        } else if(meta instanceof MapMeta) {
+            MapMeta map = (MapMeta)meta;
+            ConfigurationSection info = config.createSection("information");
+            info.set("scaling",map.isScaling());
```
Comment (CryptoMorin) [inline src/main/java/com/cryptomorin/xseries/XItemStack.java:257]: Same as above ^

**Review by CryptoMorin — CHANGES_REQUESTED** [review]

File: src/main/java/com/cryptomorin/xseries/XItemStack.java (hunk 1 above)
Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:113]: If the items have colored display names or lores, the method **ChatColor.stripColor** will ommit the colors.

File: src/main/java/com/cryptomorin/xseries/XItemStack.java (hunk 2 above)
Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:470]: https://github.com/CryptoMorin/XSeries/blob/7e8144aa7788641e20f02b8ae5e34e88b1abe88c/src/main/java/com/cryptomorin/xseries/XItemStack.java#L223

File: src/main/java/com/cryptomorin/xseries/XMaterial.java (hunk 3 above)
Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XMaterial.java:1590]: This change was made because in legacy versions,  the map is saved as "MAP" and at the moment of deserializing it will take the form of an empty map instead of a filled map (Ik, it's kind of confusing but i've already tested it and it is working as it should for legacy and newer versions), You can try yourself with the current version and will see the error i'm refering.

File: src/main/java/com/cryptomorin/xseries/XItemStack.java (hunk 4 above)
Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:244]: I think it's a good way to organize it, if you don't like it we can change it.

File: src/main/java/com/cryptomorin/xseries/XItemStack.java (hunk 5 above)
Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:257]: This is how it looks with the organization i'm proposing.
organization

File: src/main/java/com/cryptomorin/xseries/XItemStack.java (hunk 1 above)
Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:113]: i moved it to organize it in a better form.

File: src/main/java/com/cryptomorin/xseries/XItemStack.java (hunk 2 above)
Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:470]: the colors section doesn't even exist.

File: src/main/java/com/cryptomorin/xseries/XItemStack.java (hunk 1 above)
Comment (CryptoMorin) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:113]: Oh you're right, I totally forget that. I thought it's going to reverse them back to `&` good catch!

File: src/main/java/com/cryptomorin/xseries/XItemStack.java (hunk 2 above)
Comment (CryptoMorin) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:470]: I see. I think it'd be better to change `base-colors` to `colors` instead? I'm trying to reuse simple names across multiple meta types for the sake of simplicity.

Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:470]: The thing is that are two types of colors **base-colors** and **fade-colors**, if we change it then we couldn't differentiate them.

Comment (CryptoMorin) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:470]: I guess you're right. How about moving them to their own section like you did with the other meta?
```yml
colors:
  base: []
  fade: []
```

Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:470]: Sure that would be nice.

File: src/main/java/com/cryptomorin/xseries/XItemStack.java (hunk 4 above)
Comment (CryptoMorin) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:244]: Okay, we should use something that's more obvious and shorter like `book`

File: src/main/java/com/cryptomorin/xseries/XItemStack.java (hunk 5 above)
Comment (CryptoMorin) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:257]: `information` -> `map`
The `view` section is redundant imo. Just move it all to `map`
I think it'd be more convenient to use `center: <x>, <y>` format, but since that's not a common pattern in the config, it'd be better to just move them to their own section to avoid confusion.
```yml
center:
  x: 1
  y: 0
```

File: src/main/java/com/cryptomorin/xseries/XItemStack.java (hunk 4 above)
Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:245]: alright np.

File: src/main/java/com/cryptomorin/xseries/XMaterial.java (hunk 3 above)
Comment (CryptoMorin) [reply src/main/java/com/cryptomorin/xseries/XMaterial.java:1590]: Yeah, I'll need some time to test this. This material is a really special exception and it's really confusing to deal with like you mentioned.

File: src/main/java/com/cryptomorin/xseries/XItemStack.java (hunk 5 above)
Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:258]: The MapView is another object taken from MapMeta class, don't you think is good to separate them?

Comment (CryptoMorin) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:258]: Well, sure. Maybe add `setColor` and `setLocationName` support as well. [MapMeta](https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/inventory/meta/MapMeta)

Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XItemStack.java:258]: Perfect.

File: src/main/java/com/cryptomorin/xseries/XMaterial.java (hunk 3 above)
Comment (CryptoMorin) [reply src/main/java/com/cryptomorin/xseries/XMaterial.java:1590]: Can you move this to a new separate PR so it's easier to track? Thanks

Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XMaterial.java:1590]: How can i do that?, sorry i kind of new in Git & GitHub.

Comment (CryptoMorin) [reply src/main/java/com/cryptomorin/xseries/XMaterial.java:1590]: Open Git Bash in your fork's root directory and do this [this](https://stackoverflow.com/questions/39459467/)

Comment (DeadSilenceIV) [reply src/main/java/com/cryptomorin/xseries/XMaterial.java]: Done.

**Comment by DeadSilenceIV** [comment]
All the requested changes were done.

**Comment by CryptoMorin** [comment]
Looks great! Thank you so much ^^
